### PR TITLE
Mouse picking

### LINKED ISSUE #170: Mouse picking
Implement mouse picking in order to be able to select entities in the scene without going through the tree.

### PR DESCRIPTION
**Pull request by SwedishSock**
Enables you to click on entities in the editor to bring up their 'edit' menus. Beware that objects for which ABB you're inside all the time (such as 'cave') needs to be furthest down on the entity list.

fixes #170

### CODE REVIEW

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [main.cpp#L27](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/a8fba3521ae72a62f360822111f26877155cddd3/src/Editor/main.cpp#L27): Return value of function freopen() is not used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AignoredReturnValue)
1. ![MINOR][MINOR] [main.cpp#L41](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/a8fba3521ae72a62f360822111f26877155cddd3/src/Editor/main.cpp#L41): Variable 'vrSystem' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Editor/Util/MousePicking.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+#include <glm\gtx\transform.hpp>
+#include "ImGui/Theme.hpp"
+
+#include <imgui.h>
+
+MousePicking::MousePicking() {
+    /// Fake constructor.
+}
+
+void MousePicking::CreateMousePicker(Entity * cam, glm::mat4 projection) {
```
Comment (Chainsawkitten) [inline src/Editor/Util/MousePicking.cpp]: Take a const glm::mat4&

File: src/Editor/Util/MousePicking.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+MousePicking::~MousePicking() {
+    /// Deconstructor.
+}
+
+glm::vec3 MousePicking::GetCurrentRay() {
+    return this->currentRay;
+}
+
+void MousePicking::UpdateProjectionMatrix(glm::mat4 projection) {
```
Comment (Chainsawkitten) [inline src/Editor/Util/MousePicking.cpp]: Take a const glm::mat4&

File: src/Editor/Util/MousePicking.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+glm::vec2 MousePicking::GetNDC(double mouseX, double mouseY) {
+    double x = (2.0f*mouseX / MainWindow::GetInstance()->GetSize().x - 1.0f);
+    double y = 1.0f - (2.0f*mouseY / MainWindow::GetInstance()->GetSize().y);
+
+    return glm::vec2(x, y);
+}
+
+glm::vec4 MousePicking::ConvertEyeCoords(glm::vec4 clipSpaceCoordinates) {
```
Comment (Chainsawkitten) [inline src/Editor/Util/MousePicking.cpp]: Take a const glm::vec4&

File: src/Editor/Util/MousePicking.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    glm::mat4 invertProjection = glm::inverse(pMatrix);
+    glm::vec4 eyeCoords = invertProjection * clipSpaceCoordinates;
+
+    return glm::vec4(eyeCoords.x, eyeCoords.y, -1.0f, 0.0f);
+}
+
+glm::vec3 MousePicking::ConvertWorldCoords(glm::vec4 eyeCoords) {
```
Comment (Chainsawkitten) [inline src/Editor/Util/MousePicking.cpp]: Take a const glm::vec4&

File: src/Editor/Util/MousePicking.hpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+/// Use mouse cursor to select objects in editor.
+class MousePicking {
+
+public:
+
+    /// Fake constructor.
+    MousePicking();
+
+    /// Constructor.
```
Comment (Chainsawkitten) [inline src/Editor/Util/MousePicking.hpp]: Not a constructor.

File: src/Editor/Util/MousePicking.hpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include "Engine\Entity\Entity.hpp"
+
+/// Use mouse cursor to select objects in editor.
+class MousePicking {
+
+public:
```
Comment (Chainsawkitten) [inline src/Editor/Util/MousePicking.hpp]: Modifiers should be indented.

File: src/Editor/Util/MousePicking.hpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+    /// Fake constructor.
+    MousePicking();
+
+    /// Constructor.
+    void CreateMousePicker(Entity * cam, glm::mat4 projection);
```
Comment (Chainsawkitten) [inline src/Editor/Util/MousePicking.hpp]: Document parameters

File: src/Editor/Util/MousePicking.hpp [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    /// Destructor.
+    ~MousePicking();
+
+    /// Return the vector for current ray.
+    /**
+    * @return A vec3 with values from current ray in world space. Normalized.
+    */
+    glm::vec3 GetCurrentRay();
+    void UpdateProjectionMatrix(glm::mat4 projection);
```
Comment (Chainsawkitten) [inline src/Editor/Util/MousePicking.hpp]: Document method

File: src/Editor/Util/MousePicking.hpp [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+    /// Update the view matrix and recalculates the ray.
+    void Update();
+
+private:
+
+    /// Where current ray direction points in world space.
```
Comment (Chainsawkitten) [inline src/Editor/Util/MousePicking.hpp]: Private members shouldn't have doxygen comments.

File: src/Editor/Util/RayIntersection.hpp [hunk 10]
```diff
@@ -0,0 +1,23 @@
+#pragma
+#include <Engine/Geometry/AssetFileHandler.hpp>
+#include <Engine/Geometry/Model.hpp>
+
+class RayIntersection {
+
+public:
```
Comment (Chainsawkitten) [inline src/Editor/Util/RayIntersection.hpp]: Modifiers should be indented.

File: src/Editor/Util/RayIntersection.hpp [hunk 11]
```diff
@@ -0,0 +1,23 @@
+#pragma
+#include <Engine/Geometry/AssetFileHandler.hpp>
+#include <Engine/Geometry/Model.hpp>
+
+class RayIntersection {
```
Comment (Chainsawkitten) [inline src/Editor/Util/RayIntersection.hpp]: Document class.

File: src/Editor/Util/RayIntersection.hpp [hunk 12]
```diff
@@ -0,0 +1,23 @@
+#pragma
```
Comment (Chainsawkitten) [inline src/Editor/Util/RayIntersection.hpp]: pragma once

File: src/Editor/Util/RayIntersection.hpp [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+public:
+
+    /// Constructor.
+    RayIntersection();
+
+    /// Destructor.
+    ~RayIntersection();
+
+    bool RayOBBIntersect(glm::vec3 rayOrigin,
```
Comment (Chainsawkitten) [inline src/Editor/Util/RayIntersection.hpp]: Document parameters and return value

File: src/Editor/Util/RayIntersection.hpp [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+        glm::vec3 rayDirection,
+        Video::AxisAlignedBoundingBox meshData,
+        glm::mat4 modelMatrix,
+        float &outputDistance);
+
+private:
+
+};
```
Comment (Chainsawkitten) [inline src/Editor/Util/RayIntersection.hpp]: Missing newline at end of file

File: src/Editor/Util/RayIntersection.hpp [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    /// Destructor.
+    ~RayIntersection();
+
+    bool RayOBBIntersect(glm::vec3 rayOrigin,
+        glm::vec3 rayDirection,
+        Video::AxisAlignedBoundingBox meshData,
+        glm::mat4 modelMatrix,
+        float &outputDistance);
+
+private:
```
Comment (Chainsawkitten) [inline src/Editor/Util/RayIntersection.hpp]: Unnecessary modifier since there are no private members

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 4 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [main.cpp#L27](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/336cd36ee7742182b5f752fca46d550425b7c048/src/Editor/main.cpp#L27): Return value of function freopen() is not used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AignoredReturnValue)
1. ![MINOR][MINOR] [main.cpp#L41](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/336cd36ee7742182b5f752fca46d550425b7c048/src/Editor/main.cpp#L41): Variable 'vrSystem' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L148](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/336cd36ee7742182b5f752fca46d550425b7c048/src/Engine/Manager/SoundManager.cpp#L148): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Editor/Editor.cpp [hunk 16]
```diff
@@ -21,14 +23,15 @@
 Editor::Editor() {
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: MINOR Member variable 'Editor::selectedEntity' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)

File: src/Editor/Editor.cpp [hunk 17]
```diff
@@ -224,35 +233,62 @@ void Editor::Show(float deltaTime) {
...
             }
         }
 
+        // Mouse ray.
+        if (Input()->Pressed(InputHandler::SELECT) && !ImGui::IsMouseHoveringAnyWindow()) {
+            mousePicker.UpdateProjectionMatrix(cameraEntity->GetComponent < Component::Lens>()->GetProjection(glm::vec2(MainWindow::GetInstance()->GetSize().x, MainWindow::GetInstance()->GetSize().y)));
+            mousePicker.Update();
+
+            for (int i = 0; i < Hymn().world.GetEntities().size(); ++i) {
+                float intersectDistance = 0.0f;
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: MINOR The scope of the variable 'intersectDistance' can be reduced. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AvariableScope)

File: src/Editor/Util/RayIntersection.cpp [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+  
+    glm::vec3 worldPos = glm::vec3(modelMatrix[3].x, modelMatrix[3].y, modelMatrix[3].z);
+    glm::vec3 delta = worldPos - rayOrigin;
+
+    glm::vec3 xyzAxis[3];
+    float minValue[3] = { meshData.minVertex.x, meshData.minVertex.y, meshData.minVertex.z };
+    float maxValue[3] = { meshData.maxVertex.x, meshData.maxVertex.y, meshData.maxVertex.z };
+
+    for (int i = 0; i < 3; i++) {
+        xyzAxis[i] = glm::vec3(modelMatrix[i].x, modelMatrix[i].y, modelMatrix[i].z);
```
Comment (NinjaDanz3r) [inline src/Editor/Util/RayIntersection.cpp]: MINOR Variable 'xyzAxis' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 7 issues
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [main.cpp#L27](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/336cd36ee7742182b5f752fca46d550425b7c048/src/Editor/main.cpp#L27): Return value of function freopen() is not used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AignoredReturnValue)
1. ![MINOR][MINOR] [main.cpp#L42](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/336cd36ee7742182b5f752fca46d550425b7c048/src/Editor/main.cpp#L42): Variable 'vrSystem' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L149](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/336cd36ee7742182b5f752fca46d550425b7c048/src/Engine/Manager/SoundManager.cpp#L149): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 7 issues
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [main.cpp#L27](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/b913bfc73afd77b3b8d1d63d2c60cb395020502e/src/Editor/main.cpp#L27): Return value of function freopen() is not used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AignoredReturnValue)
1. ![MINOR][MINOR] [main.cpp#L42](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/b913bfc73afd77b3b8d1d63d2c60cb395020502e/src/Editor/main.cpp#L42): Variable 'vrSystem' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L149](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/b913bfc73afd77b3b8d1d63d2c60cb395020502e/src/Engine/Manager/SoundManager.cpp#L149): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 7 issues
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [main.cpp#L27](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/b255e1abe3f91183e118a1f0be404beaa15f4d35/src/Editor/main.cpp#L27): Return value of function freopen() is not used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AignoredReturnValue)
1. ![MINOR][MINOR] [main.cpp#L42](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/b255e1abe3f91183e118a1f0be404beaa15f4d35/src/Editor/main.cpp#L42): Variable 'vrSystem' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L149](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/b255e1abe3f91183e118a1f0be404beaa15f4d35/src/Engine/Manager/SoundManager.cpp#L149): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Editor/Util/MousePicking.hpp [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+
+        /// Return the vector for current ray.
+        /**
+        * @return A vec3 with values from current ray in world space. Normalized.
+        */
+        glm::vec3 GetCurrentRay();
+
+        /// Updates the mouse picker's projection matrix
+        /**
+         * @projection The projection matrix to update with.
```
Comment (Chainsawkitten) [inline src/Editor/Util/MousePicking.hpp]: @param projection

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 7 issues
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [main.cpp#L27](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/0c1262089d0ab47a91c1ae63f2a528c9b67ce54b/src/Editor/main.cpp#L27): Return value of function freopen() is not used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AignoredReturnValue)
1. ![MINOR][MINOR] [main.cpp#L42](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/0c1262089d0ab47a91c1ae63f2a528c9b67ce54b/src/Editor/main.cpp#L42): Variable 'vrSystem' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L149](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/0c1262089d0ab47a91c1ae63f2a528c9b67ce54b/src/Engine/Manager/SoundManager.cpp#L149): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**